quinn-proto/src/shared.rs

```diff
@@ -75,17 +75,31 @@ impl ConnectionId {
             len: bytes.len() as u8,
             bytes: [0; MAX_CID_SIZE],
         };
-        res.bytes[..bytes.len()].clone_from_slice(&bytes);
+        res.bytes[..bytes.len()].copy_from_slice(bytes);
+        res
+    }
+
+    /// Constructs cid by reading `len` bytes from a `Buf`
+    ///
+    /// Callers need to assure that `buf.remaining() >= len`
+    pub(crate) fn from_buf(buf: &mut impl Buf, len: usize) -> Self {
+        debug_assert!(len <= MAX_CID_SIZE);
+        let mut res = Self {
+            len: len as u8,
+            bytes: [0; MAX_CID_SIZE],
+        };
+        buf.copy_to_slice(&mut res[..len]);
         res
     }
 
     /// Decode from long header format
     pub(crate) fn decode_long(buf: &mut impl Buf) -> Option<Self> {
         let len = buf.get::<u8>().ok()? as usize;
-        match len > MAX_CID_SIZE || buf.remaining() < len {
-            false => Some(ConnectionId::new(&buf.copy_to_bytes(len))),
-            true => None,
+        if len > MAX_CID_SIZE || buf.remaining() < len {
```

Comment:
The style change here seems unrelated?

Reply:
Kind of. There was first more code in the `Some` branch, and I moved it out then because I know you prefer early returns instead of big if/else blocks. Then I added the delegate but kept it. I can change it if you prefer

Reply:
Makes sense. I do prefer if you change it back.